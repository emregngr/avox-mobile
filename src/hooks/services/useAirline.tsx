import { useQuery } from '@tanstack/react-query'
import { useEffect, useMemo, useState } from 'react'

import { getAllAirlines } from '@/services/airlineService'
import useLocaleStore from '@/store/locale'
import type { Airline } from '@/types/feature/airline'
import { useDebounce } from '@/utils/common/useDebounce'
import { parseFilterRange } from '@/utils/feature/parseFilterRange'

const ITEMS_PER_PAGE = 20

export function useAirline() {
  const { selectedLocale } = useLocaleStore()
  const [searchTerm, setSearchTerm] = useState<string>('')
  const [filters, setFilters] = useState<any>({})
  const [page, setPage] = useState<number>(1)
  const [isSearchLoading, setIsSearchLoading] = useState<boolean>(false)
  const debouncedSearch = useDebounce<string>(searchTerm, 500)

  const {
    data: allAirlinesData,
    error,
    isLoading,
  } = useQuery({
    queryFn: () => getAllAirlines(selectedLocale),
    queryKey: ['allAirlines', selectedLocale],
    staleTime: 5 * 60 * 1000,
  })

  const applyFilters = (airlines: Airline[]): Airline[] => {
    if (!airlines) return []

    return airlines.filter(airline => {
      if (!airline) return false

      if (filters?.region) {
        const airlineRegion = airline?.operations?.region?.toLowerCase()
        const filterRegion = filters?.region?.toLowerCase()

        if (airlineRegion !== filterRegion) {
          return false
        }
      }

      if (filters?.businessType) {
        const businessType = airline?.operations?.businessType?.toLowerCase()
        const filterBusinessType = filters?.businessType?.toLowerCase()

        if (businessType !== filterBusinessType) {
          return false
        }
      }

      if (filters?.businessModel) {
        const businessModel = airline?.operations?.businessModel?.toLowerCase()
        const filterBusinessModel = filters?.businessModel?.toLowerCase()

        if (businessModel !== filterBusinessModel) {
          return false
        }
      }

      if (filters?.alliance) {
        const alliance = airline?.operations?.alliance?.toLowerCase()
        const filterAlliance = filters?.alliance?.toLowerCase()

        if (alliance !== filterAlliance) {
          return false
        }
      }

      if (filters?.minSkytraxRating && airline?.operations?.skytraxRating) {
        if (airline?.operations?.skytraxRating < filters?.minSkytraxRating) {
          return false
        }
      }

      const airlineRangeChecks: {
        filterKey: keyof typeof filters
        isYear?: boolean
        jsonPath: (a: Airline) => number | undefined | null
      }[] = [
        {
          filterKey: 'foundingYear',
          isYear: true,
          jsonPath: a =>
            a.companyInfo?.foundingYear
              ? parseInt(String(a?.companyInfo?.foundingYear), 10)
              : undefined,
        },
        { filterKey: 'passengerCapacity', jsonPath: a => a?.companyInfo?.passengerCapacity },
        { filterKey: 'employeeCount', jsonPath: a => a?.companyInfo?.employeeCount },
        { filterKey: 'destinationCount', jsonPath: a => a?.network?.destinationCount },
        { filterKey: 'destinationCountries', jsonPath: a => a?.network?.destinationCountries },
        { filterKey: 'domesticConnections', jsonPath: a => a?.network?.domesticConnections },
        {
          filterKey: 'internationalConnections',
          jsonPath: a => a?.network?.internationalConnections,
        },
        { filterKey: 'totalAirplane', jsonPath: a => a?.fleet?.totalAirplane },
        { filterKey: 'averageAgeYears', jsonPath: a => a?.fleet?.averageAgeYears },
        { filterKey: 'airplaneTypeCount', jsonPath: a => a?.fleet?.airplaneTypeCount },
      ]

      for (const check of airlineRangeChecks) {
        const filterValueStr = filters?.[check.filterKey] as string
        if (filterValueStr) {
          const parsed = parseFilterRange(filterValueStr)
          if (parsed) {
            let itemValue = check?.jsonPath(airline)
            if (itemValue === undefined || itemValue === null || isNaN(Number(itemValue))) {
              return false
            }
            itemValue = Number(itemValue)
            if (itemValue < parsed.min || itemValue > parsed.max) {
              return false
            }
          }
        }
      }
      return true
    })
  }

  const filteredAirlines = useMemo(() => {
    if (!allAirlinesData) return []

    let filtered = allAirlinesData

    if (debouncedSearch.trim()) {
      const searchLower = debouncedSearch.toLowerCase().trim()

      const exactMatches = filtered.filter(
        airline =>
          airline.iataCode?.toLowerCase() === searchLower ||
          airline.icaoCode?.toLowerCase() === searchLower,
      )

      if (exactMatches.length > 0) {
        filtered = exactMatches
      } else {
        filtered = filtered.filter(airline => {
          const name = airline?.name?.toLowerCase() || ''
          const city = airline?.operations?.hub?.city?.toLowerCase() || ''
          const country = airline?.operations?.country?.toLowerCase() || ''
          const iata = airline?.iataCode?.toLowerCase() || ''
          const icao = airline?.icaoCode?.toLowerCase() || ''

          return (
            name.includes(searchLower) ||
            city.includes(searchLower) ||
            country.includes(searchLower) ||
            iata.includes(searchLower) ||
            icao.includes(searchLower)
          )
        })
      }
    }

    return applyFilters(filtered)
  }, [allAirlinesData, debouncedSearch, filters])

  const paginatedAirlines = useMemo(() => {
    if (!filteredAirlines) return []
    return filteredAirlines.slice(0, page * ITEMS_PER_PAGE)
  }, [filteredAirlines, page])

  const hasMore = paginatedAirlines.length < filteredAirlines.length

  const loadMore = () => {
    if (hasMore && !isLoading) {
      setPage(prev => prev + 1)
    }
  }

  useEffect(() => {
    if (searchTerm !== debouncedSearch) {
      setIsSearchLoading(true)
    } else {
      setTimeout(() => setIsSearchLoading(false), 300)
    }
  }, [searchTerm, debouncedSearch])

  useEffect(() => {
    setPage(1)
  }, [debouncedSearch, filters])

  return {
    error,
    filteredCount: filteredAirlines?.length,
    filters,
    hasMore,
    isLoading,
    isSearchLoading,
    loadMore,
    paginatedAirlines,
    searchTerm,
    setFilters,
    setSearchTerm,
  }
}
