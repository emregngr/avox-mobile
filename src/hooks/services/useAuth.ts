import { getApp } from '@react-native-firebase/app'
import type { FirebaseAuthTypes } from '@react-native-firebase/auth'
import { getAuth, getIdToken, onAuthStateChanged } from '@react-native-firebase/auth'
import type { QueryClient } from '@tanstack/react-query'
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import { router } from 'expo-router'
import { Alert } from 'react-native'
import Toast from 'react-native-toast-message'

import { getLocale } from '@/locales/i18next'
import {
  handleLogout,
  sendPasswordResetLink,
  signInWithApple,
  signInWithEmail,
  signInWithGoogle,
  signUpWithEmail,
} from '@/services/authService'
import { login, logout, register, social } from '@/store/auth'
import useThemeStore from '@/store/theme'
import type { AuthCredentials, LoginCredentials, RegisterCredentials } from '@/types/feature/auth'
import { Logger } from '@/utils/common/logger'

const app = getApp()
const auth = getAuth(app)

const handleAuthSuccess = async <T extends AuthCredentials>(
  userCredential: FirebaseAuthTypes.UserCredential,
  authFunction: (data: T & { token: string }) => Promise<void>,
  authData: T,
  queryClient: QueryClient,
) => {
  try {
    const token = await getIdToken(userCredential.user, true)
    await authFunction({ ...authData, token })
    queryClient.invalidateQueries({ queryKey: ['user'] })
  } catch (error) {
    Logger.breadcrumb('Failed to handle auth success', 'error', error as Error)
  }
}

export const useUser = () =>
  useQuery({
    queryFn: () =>
      new Promise(resolve => {
        const unsubscribe = onAuthStateChanged(auth, userState => {
          unsubscribe()
          resolve(userState)
        })
      }),
    queryKey: ['user'],
    staleTime: Infinity,
  })

export const useEmailRegister = () => {
  const queryClient = useQueryClient()
  const { selectedTheme } = useThemeStore()

  return useMutation({
    mutationFn: async ({ email, firstName, lastName, password }: RegisterCredentials) =>
      await signUpWithEmail({ email, firstName, lastName, password }),
    onError: () =>
      Alert.alert(
        getLocale('error'),
        getLocale('emailRegisterFailed'),
        [
          {
            text: getLocale('ok'),
          },
        ],
        {
          userInterfaceStyle: selectedTheme,
        },
      ),
    onSuccess: async (userCredential, variables) => {
      await handleAuthSuccess(
        userCredential,
        register,
        {
          email: variables.email,
          firstName: variables.firstName,
          lastName: variables.lastName,
          password: variables.password,
          token: '',
        },
        queryClient,
      )
    },
  })
}

export const useEmailLogin = () => {
  const queryClient = useQueryClient()
  const { selectedTheme } = useThemeStore()

  return useMutation({
    mutationFn: async ({ email, password }: LoginCredentials) =>
      await signInWithEmail({ email, password }),
    onError: () =>
      Alert.alert(
        getLocale('error'),
        getLocale('emailLoginFailed'),
        [
          {
            text: getLocale('ok'),
          },
        ],
        {
          userInterfaceStyle: selectedTheme,
        },
      ),
    onSuccess: async (userCredential, variables) => {
      await handleAuthSuccess(
        userCredential,
        login,
        {
          email: variables.email,
          password: variables.password,
        },
        queryClient,
      )
    },
  })
}

const useSocialLogin = (
  signInMethod: () => Promise<FirebaseAuthTypes.UserCredential>,
  provider: 'google' | 'apple',
) => {
  const queryClient = useQueryClient()
  const { selectedTheme } = useThemeStore()

  return useMutation({
    mutationFn: async () => await signInMethod(),
    onError: () =>
      Alert.alert(
        getLocale('error'),
        getLocale('socialLoginFailed'),
        [
          {
            text: getLocale('ok'),
          },
        ],
        {
          userInterfaceStyle: selectedTheme,
        },
      ),
    onSuccess: async userCredential => {
      await handleAuthSuccess(userCredential, social, { provider }, queryClient)
    },
  })
}

export const useGoogleLogin = () => useSocialLogin(signInWithGoogle, 'google')

export const useAppleLogin = () => useSocialLogin(signInWithApple, 'apple')

export const useLogout = () => {
  const queryClient = useQueryClient()
  const { selectedTheme } = useThemeStore()

  return useMutation({
    mutationFn: async () => await handleLogout(),
    onError: () => {
      Alert.alert(
        getLocale('error'),
        getLocale('logoutFailed'),
        [
          {
            text: getLocale('ok'),
          },
        ],
        {
          userInterfaceStyle: selectedTheme,
        },
      )
    },
    onSuccess: async () => {
      await logout()
      queryClient.setQueryData(['user'], null)
      queryClient.removeQueries()
    },
  })
}

export const useForgotPassword = () => {
  const { selectedTheme } = useThemeStore()

  return useMutation({
    mutationFn: sendPasswordResetLink,
    onError: () =>
      Alert.alert(
        getLocale('error'),
        getLocale('passwordResetLinkSentMessageFailed'),
        [
          {
            text: getLocale('ok'),
          },
        ],
        {
          userInterfaceStyle: selectedTheme,
        },
      ),
    onSuccess: () => {
      Toast.show({
        text1: getLocale('successful'),
        text2: getLocale('passwordResetLinkSentMessage'),
        type: 'success',
      })
      router.back()
    },
  })
}
