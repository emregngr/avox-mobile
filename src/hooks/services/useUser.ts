import { getApp } from '@react-native-firebase/app'
import { getAuth } from '@react-native-firebase/auth'
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import { router } from 'expo-router'
import { Alert } from 'react-native'
import Toast from 'react-native-toast-message'

import { getLocale } from '@/locales/i18next'
import {
  addUserPassword,
  changeUserPassword,
  deleteUserAccount,
  getUser,
  registerDevice,
  registerDeviceToUser,
  updateUser,
} from '@/services/userService'
import { setIsAuthenticated } from '@/store/auth'
import useThemeStore from '@/store/theme'
import { deleteUser } from '@/store/user'
import type { AddPasswordCredentials, ChangePasswordCredentials } from '@/types/feature/password'
import type { ProfileData } from '@/types/feature/user'

const app = getApp()
const auth = getAuth(app)

export const useGetUser = () => {
  const userId = auth?.currentUser?.uid
  return useQuery({
    enabled: !!userId,
    queryFn: getUser,
    queryKey: ['userProfile', userId],
  })
}

export const useUpdateUser = () => {
  const queryClient = useQueryClient()
  const userId = auth?.currentUser?.uid
  const { selectedTheme } = useThemeStore()

  return useMutation<any, Error, ProfileData, { previousProfile: any }>({
    mutationFn: async (profileData: ProfileData) => await updateUser(profileData),
    onError: (error, newProfileData, context) => {
      queryClient.setQueryData(['userProfile', userId], context?.previousProfile)
      Alert.alert(
        getLocale('error'),
        getLocale('profileUpdateFailed'),
        [
          {
            text: getLocale('ok'),
          },
        ],
        {
          userInterfaceStyle: selectedTheme,
        },
      )
    },
    onMutate: async newProfileData => {
      await queryClient.cancelQueries({ queryKey: ['userProfile', userId] })
      const previousProfile = queryClient.getQueryData(['userProfile', userId])
      queryClient.setQueryData(['userProfile', userId], (oldData: any) => ({
        ...oldData,
        ...newProfileData,
        displayName: `${newProfileData?.firstName} ${newProfileData?.lastName}`,
      }))
      return { previousProfile }
    },
    onSettled: () => {
      queryClient.invalidateQueries({ queryKey: ['userProfile', userId] })
    },
    onSuccess: () => {
      Toast.show({
        text1: getLocale('successful'),
        text2: getLocale('profileUpdatedSuccessfully'),
        type: 'success',
      })
      router.back()
    },
  })
}

export const useChangePassword = () => {
  const { selectedTheme } = useThemeStore()

  return useMutation({
    mutationFn: async (credentials: ChangePasswordCredentials) =>
      await changeUserPassword(credentials),
    onError: () => {
      Alert.alert(
        getLocale('error'),
        getLocale('passwordChangeFailed'),
        [
          {
            text: getLocale('ok'),
          },
        ],
        {
          userInterfaceStyle: selectedTheme,
        },
      )
    },
    onSuccess: () => {
      Toast.show({
        text1: getLocale('successful'),
        text2: getLocale('passwordChangedSuccessfully'),
        type: 'success',
      })
      router.back()
    },
  })
}

export const useAddPassword = () => {
  const { selectedTheme } = useThemeStore()

  return useMutation({
    mutationFn: async (credentials: AddPasswordCredentials) => await addUserPassword(credentials),
    onError: () => {
      Alert.alert(
        getLocale('error'),
        getLocale('passwordAddFailed'),
        [
          {
            text: getLocale('ok'),
          },
        ],
        {
          userInterfaceStyle: selectedTheme,
        },
      )
    },
    onSuccess: () => {
      Toast.show({
        text1: getLocale('successful'),
        text2: getLocale('passwordAddedSuccessfully'),
        type: 'success',
      })
      router.back()
    },
  })
}

export const useDeleteUser = () => {
  const queryClient = useQueryClient()

  return useMutation({
    mutationFn: deleteUserAccount,
    onError: () => {
      Alert.alert(getLocale('error'), getLocale('accountDeleteFailed'), [
        {
          text: getLocale('ok'),
        },
      ])
    },
    onSuccess: async () => {
      await deleteUser()
      setIsAuthenticated(false)
      queryClient.clear()
      Toast.show({
        text1: getLocale('successful'),
        text2: getLocale('accountDeletedSuccessfully'),
        type: 'success',
      })
    },
  })
}

export const useRegisterDevice = () =>
  useMutation({
    mutationFn: async () => await registerDevice(),
  })

export const useRegisterDeviceToUser = () =>
  useMutation({
    mutationFn: async () => await registerDeviceToUser(),
  })
