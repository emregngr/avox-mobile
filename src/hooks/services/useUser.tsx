import firebase from '@react-native-firebase/app'
import { getAuth } from '@react-native-firebase/auth'
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import { router } from 'expo-router'
import { Alert } from 'react-native'
import Toast from 'react-native-toast-message'

import { getLocale } from '@/locales/i18next'
import {
  addUserPassword,
  changeUserPassword,
  deleteUserAccount,
  getUser,
  registerDevice,
  registerDeviceToUser,
  updateUser,
} from '@/services/userService'
import { setIsAuthenticated } from '@/store/auth'
import { deleteUser } from '@/store/user'
import type { AddPasswordCredentials, ChangePasswordCredentials } from '@/types/feature/password'
import type { ProfileData } from '@/types/feature/user'

const auth = getAuth(firebase.app())

export const useGetUser = () => {
  const userId = auth?.currentUser?.uid
  return useQuery({
    enabled: !!userId,
    queryFn: getUser,
    queryKey: ['userProfile', userId],
  })
}

export const useUpdateUser = () => {
  const queryClient = useQueryClient()
  const userId = auth?.currentUser?.uid

  return useMutation<any, Error, ProfileData, { previousProfile: any }>({
    mutationFn: (profileData: ProfileData) => updateUser(profileData),
    onError: (error, newProfileData, context) => {
      queryClient.setQueryData(['userProfile', userId], context?.previousProfile)
      Alert.alert(getLocale('error'), getLocale('profileUpdateFailed'), [
        {
          text: getLocale('ok'),
        },
      ])
    },
    onMutate: async newProfileData => {
      await queryClient.cancelQueries({ queryKey: ['userProfile', userId] })
      const previousProfile = queryClient.getQueryData(['userProfile', userId])
      queryClient.setQueryData(['userProfile', userId], (oldData: any) => ({
        ...oldData,
        ...newProfileData,
        displayName: `${newProfileData?.firstName} ${newProfileData?.lastName}`,
      }))
      return { previousProfile }
    },
    onSettled: () => {
      queryClient.invalidateQueries({ queryKey: ['userProfile', userId] })
    },
    onSuccess: () => {
      Toast.show({
        text1: getLocale('successful'),
        text2: getLocale('profileUpdatedSuccessfully'),
        type: 'success',
      })
      router.back()
    },
  })
}

export const useChangePassword = () =>
  useMutation({
    mutationFn: (credentials: ChangePasswordCredentials) => changeUserPassword(credentials),
    onError: () => {
      Alert.alert(getLocale('error'), getLocale('passwordChangeFailed'), [
        {
          text: getLocale('ok'),
        },
      ])
    },
    onSuccess: () => {
      Toast.show({
        text1: getLocale('successful'),
        text2: getLocale('passwordChangedSuccessfully'),
        type: 'success',
      })
      router.back()
    },
  })

export const useAddPassword = () =>
  useMutation({
    mutationFn: (credentials: AddPasswordCredentials) => addUserPassword(credentials),
    onError: () => {
      Alert.alert(getLocale('error'), getLocale('passwordAddFailed'), [
        {
          text: getLocale('ok'),
        },
      ])
    },
    onSuccess: () => {
      Toast.show({
        text1: getLocale('successful'),
        text2: getLocale('passwordAddedSuccessfully'),
        type: 'success',
      })
      router.back()
    },
  })

export const useDeleteUser = () => {
  const queryClient = useQueryClient()

  return useMutation({
    mutationFn: deleteUserAccount,
    onError: () => {
      Alert.alert(getLocale('error'), getLocale('accountDeleteFailed'), [
        {
          text: getLocale('ok'),
        },
      ])
    },
    onSuccess: () => {
      deleteUser()
      setIsAuthenticated(false)
      queryClient.clear()
      Toast.show({
        text1: getLocale('successful'),
        text2: getLocale('accountDeletedSuccessfully'),
        type: 'success',
      })
    },
  })
}

export const useRegisterDevice = () =>
  useMutation({
    mutationFn: registerDevice,
  })

export const useRegisterDeviceToUser = () =>
  useMutation({
    mutationFn: registerDeviceToUser,
  })
