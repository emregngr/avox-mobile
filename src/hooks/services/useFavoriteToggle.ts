import { router } from 'expo-router'
import { useCallback, useMemo } from 'react'

import { useAddFavorite, useIsFavorite, useRemoveFavorite } from '@/hooks/services/useFavorite'
import useAuthStore from '@/store/auth'
import type { FavoriteItemType } from '@/types/feature/favorite'
import { Logger } from '@/utils/common/logger'

export function useFavoriteToggle({ id, type }: FavoriteItemType) {
  const { isAuthenticated } = useAuthStore()

  const { isPending: isAddFavoritePending, mutateAsync: addFavoriteMutation } = useAddFavorite()
  const { isPending: isRemoveFavoritePending, mutateAsync: removeFavoriteMutation } =
    useRemoveFavorite()

  const isFavorite = useIsFavorite({ id, type })

  const isPending = useMemo(
    () => isAddFavoritePending || isRemoveFavoritePending,
    [isAddFavoritePending, isRemoveFavoritePending],
  )

  const handleFavoritePress = useCallback(async (): Promise<void> => {
    if (!isAuthenticated) {
      router.replace('/auth')
      return
    }

    if (isPending) return

    try {
      if (isFavorite) {
        await removeFavoriteMutation({ id, type })
      } else {
        await addFavoriteMutation({ id, type })
      }
    } catch (error) {
      Logger.breadcrumb('Failed to toggle favorite', 'error', error as Error)
    }
  }, [
    isAuthenticated,
    isPending,
    isFavorite,
    id,
    type,
    addFavoriteMutation,
    removeFavoriteMutation,
  ])

  return { handleFavoritePress, isFavorite, isPending }
}
