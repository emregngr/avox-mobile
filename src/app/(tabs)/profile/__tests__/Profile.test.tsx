import { fireEvent, render } from '@testing-library/react-native'
import { router } from 'expo-router'
import type React from 'react'
import type { ReactNode } from 'react'
import { Alert } from 'react-native'
import { SafeAreaProvider } from 'react-native-safe-area-context'

import Profile from '@/app/(tabs)/profile'
import { useLogout } from '@/hooks/services/useAuth'
import { useGetUser } from '@/hooks/services/useUser'
import useLocaleStore from '@/store/locale'
import useThemeStore from '@/store/theme'
const { mockedAuth } = require('@react-native-firebase/auth')

jest.mock('@/locales/i18next', () => ({
  getLocale: (key: string) => key,
}))

jest.mock('@/assets/icons/settings.svg', () => 'Settings')

jest.mock('@/hooks/services/useAuth')

const mockedUseLogout = useLogout as jest.MockedFunction<typeof useLogout>

jest.mock('@/hooks/services/useUser')

const mockedUseGetUser = useGetUser as jest.MockedFunction<typeof useGetUser>

jest.mock('@/store/locale')

const mockedUseLocaleStore = useLocaleStore as jest.MockedFunction<typeof useLocaleStore>

jest.mock('@/store/theme')

const mockedUseThemeStore = useThemeStore as jest.MockedFunction<typeof useThemeStore>

jest.mock('@/components/common', () => {
  const { TouchableOpacity, Text, View } = require('react-native')

  return {
    Header: ({
      title,
      rightIconOnPress,
      rightIcon,
    }: {
      title: string
      rightIconOnPress?: () => void
      rightIcon?: ReactNode
    }) => (
      <View testID="header">
        <Text testID="header-title">{title}</Text>
        {rightIconOnPress ? (
          <TouchableOpacity onPress={rightIconOnPress} testID="settings-button">
            {rightIcon}
          </TouchableOpacity>
        ) : null}
      </View>
    ),

    ProfileItem: ({ label, onPress }: { label: string; onPress: () => void }) => (
      <TouchableOpacity onPress={onPress} testID={`item-${label}`}>
        <Text>{label}</Text>
      </TouchableOpacity>
    ),

    SafeLayout: ({ children }: { children: ReactNode }) => (
      <View testID="safe-layout">{children}</View>
    ),

    ThemedText: ({ children, ...props }: { children: string }) => (
      <Text {...props}>{children}</Text>
    ),
  }
})

const mockedUserWithPassword = {
  uid: 'user123',
  displayName: 'John Doe',
  photoURL: 'http://example.com/photo.png',
  providerData: [{ providerId: 'password' }],
}

const mockedUserWithoutPassword = {
  uid: 'user456',
  displayName: 'Jane Smith',
  photoURL: null,
  providerData: [{ providerId: 'google.com' }],
}

let mockedAlert: jest.SpyInstance

const renderWithSafeAreaProvider = (component: ReactNode) =>
  render(
    <SafeAreaProvider
      initialMetrics={{
        insets: { top: 0, left: 0, right: 0, bottom: 0 },
        frame: { x: 0, y: 0, width: 0, height: 0 },
      }}
    >
      {component}
    </SafeAreaProvider>,
  )

beforeEach(() => {
  mockedUseLocaleStore.mockReturnValue({ selectedLocale: 'en' })
  mockedUseThemeStore.mockReturnValue({ selectedTheme: 'light' })

  mockedAuth.currentUser = null

  mockedUseLogout.mockReturnValue({ mutateAsync: jest.fn() } as any)

  mockedUseGetUser.mockReturnValue({ data: null } as any)

  mockedAlert = jest.spyOn(Alert, 'alert').mockImplementation(() => {})
})

describe('Profile Screen', () => {
  describe('When user is logged out', () => {
    it('should render the "Sign In or Register" option', () => {
      const { getByTestId, queryByTestId } = renderWithSafeAreaProvider(<Profile />)
      expect(getByTestId('item-signInOrRegister')).toBeTruthy()
      expect(queryByTestId('item-updateProfile')).toBeNull()
      expect(queryByTestId('item-logout')).toBeNull()
    })

    it('should display default name and photo', () => {
      const { getByText } = renderWithSafeAreaProvider(<Profile />)
      expect(getByText('avox')).toBeTruthy()
    })

    it('should navigate to auth screen on press', () => {
      const { getByTestId } = renderWithSafeAreaProvider(<Profile />)
      fireEvent.press(getByTestId('item-signInOrRegister'))
      expect(router.replace).toHaveBeenCalledWith({
        params: { tab: 'profile' },
        pathname: '/auth',
      })
    })
  })

  describe('When user is logged in', () => {
    it('should render user-specific  items', () => {
      mockedAuth.currentUser = mockedUserWithPassword
      mockedUseGetUser.mockReturnValue({ data: mockedUserWithPassword } as any)

      const { getByTestId, queryByTestId, getByText } = renderWithSafeAreaProvider(<Profile />)

      expect(getByText('John Doe')).toBeTruthy()
      expect(getByTestId('item-updateProfile')).toBeTruthy()
      expect(getByTestId('item-logout')).toBeTruthy()
      expect(queryByTestId('item-signInOrRegister')).toBeNull()
    })

    it('should show "Change Password" for password users', () => {
      mockedAuth.currentUser = mockedUserWithPassword
      mockedUseGetUser.mockReturnValue({ data: mockedUserWithPassword } as any)

      const { getByTestId } = renderWithSafeAreaProvider(<Profile />)
      expect(getByTestId('item-changePassword')).toBeTruthy()
    })

    it('should show "Add Password" for non-password users', () => {
      mockedAuth.currentUser = mockedUserWithoutPassword
      mockedUseGetUser.mockReturnValue({ data: mockedUserWithoutPassword } as any)

      const { getByTestId } = renderWithSafeAreaProvider(<Profile />)
      expect(getByTestId('item-addPassword')).toBeTruthy()
    })

    it('should show a confirmation alert on logout press', () => {
      mockedAuth.currentUser = mockedUserWithPassword
      mockedUseGetUser.mockReturnValue({ data: mockedUserWithPassword } as any)

      const { getByTestId } = renderWithSafeAreaProvider(<Profile />)
      fireEvent.press(getByTestId('item-logout'))

      expect(mockedAlert).toHaveBeenCalledWith(
        'warning',
        'logoutWarning',
        expect.any(Array),
        expect.any(Object),
      )
    })

    it('should call the logout mutation when confirming the alert', () => {
      mockedAuth.currentUser = mockedUserWithPassword
      mockedUseGetUser.mockReturnValue({ data: mockedUserWithPassword } as any)
      mockedAlert.mockImplementation((title, message, buttons) => {
        const yesButton = buttons.find((b: any) => b.style === 'destructive')
        if (yesButton) {
          yesButton.onPress()
        }
      })

      const { getByTestId } = renderWithSafeAreaProvider(<Profile />)
      fireEvent.press(getByTestId('item-logout'))

      expect(mockedUseLogout).toHaveBeenCalledTimes(1)
    })
  })

  describe('Navigation', () => {
    it('should navigate to settings on header icon press', () => {
      const { getByTestId } = renderWithSafeAreaProvider(<Profile />)
      fireEvent.press(getByTestId('settings-button'))
      expect(router.navigate).toHaveBeenCalledWith('/settings')
    })

    it('should navigate to choose-theme on item press', () => {
      const { getByTestId } = renderWithSafeAreaProvider(<Profile />)
      fireEvent.press(getByTestId('item-chooseTheme'))
      expect(router.navigate).toHaveBeenCalledWith('/choose-theme')
    })

    it('should navigate to choose-language on item press', () => {
      const { getByTestId } = renderWithSafeAreaProvider(<Profile />)
      fireEvent.press(getByTestId('item-chooseLanguage'))
      expect(router.navigate).toHaveBeenCalledWith('/choose-language')
    })
  })
})

describe('Profile Screen Snapshot', () => {
  it('should render the Profile screen successfully', () => {
    const { toJSON } = renderWithSafeAreaProvider(<Profile />)

    expect(toJSON()).toMatchSnapshot()
  })
})
