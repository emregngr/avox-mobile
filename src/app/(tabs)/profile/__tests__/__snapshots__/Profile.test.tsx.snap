// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Profile Screen Snapshot should render the Profile screen successfully 1`] = `
<View
  testID="safe-layout"
>
  <View
    testID="header"
  >
    <Text
      testID="header-title"
    />
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
      testID="settings-button"
    >
      <Settings
        color="#000000"
        height={24}
        width={24}
      />
    </View>
  </View>
  <RCTScrollView
    contentContainerClassName="px-4"
    contentContainerStyle={
      {
        "paddingBottom": 88,
      }
    }
    showsVerticalScrollIndicator={false}
  >
    <View>
      <View
        className="items-center mt-5"
      >
        <View
          className="w-24 h-24 rounded-full overflow-hidden border-[3px] border-onPrimary-100 justify-center items-center mb-4"
        >
          <View
            cachePolicy="memory-disk"
            contentFit="contain"
            source={
              {
                "default": "mocked-test-file",
              }
            }
            style={
              {
                "borderRadius": 40,
                "height": 80,
                "width": 80,
              }
            }
            testID="mocked-image"
            transition={0}
          />
        </View>
        <Text
          color="text-100"
          type="body2"
        >
          avox
        </Text>
      </View>
      <Text
        className="ml-4 mt-8 mb-2"
        color="text-70"
        type="body2"
      >
        profileInformation
      </Text>
      <View
        className="rounded-xl overflow-hidden bg-background-secondary"
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
          testID="item-signInOrRegister"
        >
          <Text>
            signInOrRegister
          </Text>
        </View>
      </View>
      <Text
        className="ml-4 mt-8 mb-2"
        color="text-70"
        type="body2"
      >
        accountTransactions
      </Text>
      <View
        className="rounded-xl overflow-hidden bg-background-secondary"
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
          testID="item-chooseTheme"
        >
          <Text>
            chooseTheme
          </Text>
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
          testID="item-chooseLanguage"
        >
          <Text>
            chooseLanguage
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
